packages/miniflare/src/plugins/core/inspector-proxy/index.ts

```diff
@@ -0,0 +1,269 @@
+		// versions working. Instead, we look for a browser-like `User-Agent`.
+		const userAgent = req.headers["user-agent"] ?? "";
+		const hasFileSystemAccess = !/mozilla/i.test(userAgent);
+
+		return hasFileSystemAccess;
+	}
+
+	#inspectorId = crypto.randomUUID();
+	async #handleDevToolsJsonRequest(host: string, path: string) {
+		if (path === "/json/version") {
+			return {
+				Browser: `miniflare/v${miniflareVersion}`,
+				// TODO: (someday): The DevTools protocol should match that of workerd.
+				// This could be exposed by the preview API.
+				"Protocol-Version": "1.3",
+			};
+		}
+
+		if (path === "/json" || path === "/json/list") {
+			return this.#workerRelays.map((relay) => {
+				const workerName = relay.workerName;
+				const localHost = `${host}/${workerName}`;
+				const devtoolsFrontendUrl = `https://devtools.devprod.cloudflare.dev/js_app?theme=systemPreferred&debugger=true&ws=${localHost}`;
+
+				return {
+					id: `${this.#inspectorId}-${workerName}`,
+					type: "node", // TODO: can we specify different type?
+					description: "workers",
+					webSocketDebuggerUrl: `ws://${localHost}`,
+					devtoolsFrontendUrl,
+					devtoolsFrontendUrlCompat: devtoolsFrontendUrl,
+					// Below are fields that are visible in the DevTools UI.
+					title:
+						workerName.length === 0 || this.#workerRelays.length === 1
+							? `Cloudflare Worker`
+							: `Cloudflare Worker: ${workerName}`,
+					faviconUrl: "https://workers.cloudflare.com/favicon.ico",
+					// url: "http://" + localHost, // looks unnecessary
+				};
+			});
+		}
+
+		return null;
+	}
+
+	#runtimeMessageCounter = 1e8;
+	nextCounter() {
+		return ++this.#runtimeMessageCounter;
+	}
+
+	getInspectorURL(): URL {
+		return getWebsocketURL(this.userInspectorPort);
+	}
+
+	async updateConnection(runtimeInspectorPort: number) {
+		const workerdInspectorJson = (await fetch(
+			`http://127.0.0.1:${runtimeInspectorPort}/json`
+		).then((resp) => resp.json())) as {
+			id: string;
+		}[];
+
+		this.#workerRelays = workerdInspectorJson
+			.map(({ id }) => {
+				if (!id.startsWith("core:user:")) {
+					return;
+				}
+
+				const workerName = id.replace(/^core:user:/, "");
+
+				if (!this.workerNamesToProxy.has(workerName)) {
+					return;
+				}
+
+				return new InspectorProxyRelay(
+					workerName,
+					new WebSocket(`ws://127.0.0.1:${runtimeInspectorPort}/${id}`)
+				);
+			})
+			.filter(Boolean) as InspectorProxyRelay[];
+
+		this.#runtimeConnectionEstablished.resolve();
+	}
+
+	async #waitForReady() {
+		await this.#runtimeConnectionEstablished;
+	}
+
+	get ready(): Promise<void> {
+		return this.#waitForReady();
+	}
+
+	async dispose(): Promise<void> {
+		await Promise.all(this.#workerRelays.map((relay) => relay.dispose()));
+
+		const deferredPromise = new DeferredPromise<void>();
+		this.#server.close((err) => {
+			if (err) {
+				deferredPromise.reject(err);
+			} else {
+				deferredPromise.resolve();
+			}
+		});
+		return deferredPromise;
+	}
```

Comment:
Out of interest why not use a normal Promise here?

```suggestion
		return new Promise((resolve, reject) => {
			this.#server.close((err) => err ? reject(err) : resolve());
		});
```

Reply:
yeah you're totally right 😅

I was just using deferred promises and I think I had the the pattern stuck in my head when I implemented this, thanks! 🙂